api/breaker/breaker.go

```diff
@@ -0,0 +1,436 @@
+	}
+
+	if c.TrippedPeriod <= 0 {
+		c.TrippedPeriod = defaults.TrippedPeriod
+	}
+
+	if c.RecoveryRampPeriod <= 0 {
+		c.RecoveryRampPeriod = defaults.RecoveryRampPeriod
+	}
+
+	if c.RecoveryLimit <= 0 {
+		c.RecoveryLimit = defaults.RecoveryLimit
+	}
+
+	if c.OnTripped == nil {
+		c.OnTripped = func() {}
+	}
+
+	if c.OnStandBy == nil {
+		c.OnStandBy = func() {}
+	}
+
+	if c.IsSuccessful == nil {
+		c.IsSuccessful = NonNilErrorIsSuccess
+	}
+
+	c.TrippedPeriod = utils.NewSeventhJitter()(c.TrippedPeriod)
+
+	return nil
+}
+
+// CircuitBreaker implements the circuit breaker pattern
+type CircuitBreaker struct {
+	cfg Config
+
+	rc *ratioController
+
+	mu         sync.Mutex
+	state      State
+	generation uint64
+	metrics    Metrics
+	expiry     time.Time
+}
+
+func NewNoop() *CircuitBreaker {
+	return &CircuitBreaker{
+		cfg: NoopBreakerConfig(),
+	}
+}
+
+// New returns a CircuitBreaker configured with the provided Config
+func New(cfg Config) (*CircuitBreaker, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, err
+	}
+
+	cb := CircuitBreaker{cfg: cfg}
+	cb.nextGeneration(cfg.Clock.Now())
+
+	return &cb, nil
+}
+
+// Execute calls the provided function depending on the CircuitBreaker state.
+// - StateStandby: all functions are executed.
+// - StateTripped: no functions are executed and ErrStateTripped is returned.
+// - StateRecovering: some functions are executed, some functions are not,
+//                    when not executed ErrLimitExceeded is returned.
+//
+// The CircuitBreaker state is updated according to the outcome of executing the
+// provided function and the current state. See package docs for a more detailed
+// explanation of state transitions.
+func (c *CircuitBreaker) Execute(f func() (interface{}, error)) (interface{}, error) {
+	generation, err := c.beforeExecution()
+	if err != nil {
+		return nil, err
+	}
+
+	v, err := f()
+
+	c.afterExecution(generation, v, err)
+
+	return v, err
+}
+
+// beforeExecution checks the current state to determine if a new generation
+// should be created and whether Execute is allowed to proceed.
+func (c *CircuitBreaker) beforeExecution() (uint64, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	now := c.cfg.Clock.Now()
+
+	generation, state := c.currentState(now)
+
+	switch {
+	case state == StateTripped:
+		return generation, ErrStateTripped
+	case state == StateRecovering && !c.rc.allowRequest():
+		return generation, ErrRecoveryLimitExceeded
+	}
+
+	c.metrics.execute()
+	return generation, nil
+}
+
+// afterExecution updates the CircuitBreaker state based on the outcome of
+// processing the fn in Execute.
+func (c *CircuitBreaker) afterExecution(prior uint64, v interface{}, err error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	now := c.cfg.Clock.Now()
+
+	generation, state := c.currentState(now)
+	if generation != prior {
+		return
+	}
+
+	if c.cfg.IsSuccessful(v, err) {
+		grpclog.Infof("[breaker] successful execution, %s", c.metrics.String())
```

Comment:
Do we want to use `grpclog` here instead of `logrus` ? It might get confusing with controlling log output verbosity if we use `logrus` in most places and then use `grpclog` here.

Reply:
For sure, we should definitely look to get an interface here eventually for logging.

I've used https://github.com/go-logr/logr previously, but we could easily match that interface without using the package and still allow people to use `logr` compatible loggers.